### PR TITLE
Maintain wabt patches in oak tree

### LINKED ISSUE #332: Vendor patches needed to build wabt into the oak repository
Currently we are depending on a fork of wabt including some additional commits that add a BUILD file for Bazel and other minor changes:

https://github.com/project-oak/oak/blob/2d1c17c26fda5590b643a0e89a0985d052a7fdee/WORKSPACE#L94-L98

This is hard to maintain, and in fact we are stuck at a relatively old version of wabt because of this friction.

We should:

- switch from `git_repository` to `http_archive`: https://docs.bazel.build/versions/master/repo/http.html#http_archive
- use the `build_file` attribute of `http_archive` to specify the contents of the BUILD file that is needed to build wabt, while keeping the BUILD file itself in oak
- figure out which of the additional commits in https://github.com/daviddrysdale/wabt/commits/master still need to be patched, and apply them via the `patches` attribute of `http_archive`, maintaining the patches themselves in oak
- point the dependency back to the upstream wabt repository (possibly at an old commit to start with)
- update the dependency on wabt to point to a more recent commit

### PR DESCRIPTION
**Pull request by blaxill**
Fixes #332 

- Changes Bazel rules to use http_archive pointing at https://github.com/WebAssembly/wabt
- Keeps patches in oak tree
- Updated to master wabt

### CODE REVIEW

**Comment by tiziano88** [comment]
It looks like the `run_examples` step is timing out at 30 minutes:https://pantheon.corp.google.com/cloud-build/builds/53b4a839-3849-4c26-8004-1e84e52d8981;step=7?project=oak-ci

This is because of the constraint here:

https://github.com/project-oak/oak/blob/2d1c17c26fda5590b643a0e89a0985d052a7fdee/cloudbuild.yaml#L48-L52

Feel free to increase the timeout, hopefully this will become faster again once the bazel cache is properly initialized with the new artifacts.

**Comment by blaxill** [comment]
Hmm although the test cases complete, run_examples is actually blocking indefinitely (reproduced locally) on the blocking read at https://github.com/project-oak/oak/blob/f4d5627dd46dbedd40f1915c224348fbef39c6d9/oak/server/module_invocation.cc#L167

I'm not sure why yet.

**Comment by tiziano88** [comment]
Ah, interesting, maybe something changed in the way wabt handles imported function calls?

Does it work fine if you point the http_archive rule to the old commit upstream, from which we branched?

Given the extra complexity here and the potential of introducing subtle (or not so subtle) bugs because of different semantics between versions of wabt, I suggest doing this in two steps. If anything seems off after bumping up the version of wabt we depend on, it will be easier to roll back only that, rather than having to undo all the work you have done on the Bazel rules already. WDYT?

**Comment by blaxill** [comment]
Yes I think you're right, this could possibly be a subtle issue that would be better addressed separately from the original issue- I'll switch back to the old commit.

**Comment by blaxill** [comment]
Closing in favour of  #338

File: third_party/wabt/0004-Downgrade-nearbyint-and-trunc-to-their-C-counterpart.patch [hunk 1]
```diff
@@ -0,0 +1,33 @@
+From 002cd695d6274a4b34b931e062dd769db015eebf Mon Sep 17 00:00:00 2001
+From: Tiziano Santoro <tiziano88@gmail.com>
+Date: Wed, 31 Oct 2018 12:11:29 +0000
+Subject: [PATCH 4/8] Downgrade nearbyint and trunc to their C counterparts
+
+They are not supported in Asylo yet
```
[the remark this answers is not in the document]
Comment (blaxill) [reply third_party/wabt/0004-Downgrade-nearbyint-and-trunc-to-their-C-counterpart.patch]: I'll check this out now

File: third_party/wabt/0008-Use-std-stringstream-to-build-unique-name.patch [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+diff --git a/src/binary-reader-ir.cc b/src/binary-reader-ir.cc
+index 8f8526f1..957d218c 100644
+--- src/binary-reader-ir.cc
++++ src/binary-reader-ir.cc
+@@ -21,6 +21,7 @@
+ #include <cstdarg>
+ #include <cstdint>
+ #include <cstdio>
++#include <sstream>
```
[the remark this answers is not in the document]
Comment (blaxill) [reply third_party/wabt/0008-Use-std-stringstream-to-build-unique-name.patch]: And I'll check this now

**Comment by blaxill** [comment]
Regarding patch description, these have been generated from the git commits required to patch the point the upstream wabt deviated from [this pinned wabt](https://github.com/project-oak/oak/blob/2d1c17c26fda5590b643a0e89a0985d052a7fdee/WORKSPACE#L96). I can add my own comments for clarifying history, but perhaps it's better if @tiziano88 can clarify on the patch descriptions?

In this PR I attempted to also rebase on wabt master, but due to the failing example I'm taking the more conservative route #338 of leaving the rebase on master to later.

File: third_party/wabt/0008-Use-std-stringstream-to-build-unique-name.patch (hunk 2 above)
[the remark this answers is not in the document]
Comment (blaxill) [reply third_party/wabt/0008-Use-std-stringstream-to-build-unique-name.patch]: Ah, the rebase on wabt master removed the stringstream code. In #338 this patch file still includes the stringstream changes.

File: third_party/wabt/0004-Downgrade-nearbyint-and-trunc-to-their-C-counterpart.patch (hunk 1 above)
[the remark this answers is not in the document]
Comment (blaxill) [reply third_party/wabt/0004-Downgrade-nearbyint-and-trunc-to-their-C-counterpart.patch]: It appears this isn't needed for #338 so I'll remove it there (examples/tests pass locally).

**Pull request closed without merging**